Model archiver API

### 🚀 The feature

A python API for model archiver

### Motivation, pitch

It'd be nice to be able to archive a model  directly from the python application, without having to run a bash command with a subshell or others.

### Alternatives

I've used a script like this which is almost identical to what the entry point of `torch-model-archiver` is. The only addition is a pydantic model as config instead of arg parse. 

The application then imports `ModelArchiverConfig` and `generate_model_archive` to build a new mar file.

One of the limitations here is not being able to pass a custom handler in a clean way when it's part of the library/application.


```python
from typing import Optional, Literal, Union
import os
import logging
import shutil
from pathlib import Path

from pydantic import BaseModel
from model_archiver.model_archiver_error import ModelArchiverError
from model_archiver.model_packaging_utils import ModelExportUtils
from model_archiver.manifest_components.manifest import RuntimeType


class ModelArchiverConfig(BaseModel):
    model_name: str
    serialized_file: Optional[str] = None
    model_file: Optional[str] = None
    handler: str
    extra_files: Optional[str] = None
    runtime: str = RuntimeType.PYTHON.value
    export_path: Union[str, Path] = os.getcwd()
    archive_format: Literal["default", "tgz", "no-archive"] = "default"
    force: bool = False
    version: str
    requirements_file: Optional[str] = None


def _package_model(config: ModelArchiverConfig, manifest: str):
    """
    Internal helper for the exporting model command line interface.
    """
    model_file = config.model_file
    serialized_file = config.serialized_file
    model_name = config.model_name
    handler = config.handler
    extra_files = config.extra_files
    export_file_path = config.export_path
    requirements_file = config.requirements_file

    try:
        ModelExportUtils.validate_inputs(model_name, export_file_path)
        # Step 1 : Check if .mar already exists with the given model name
        export_file_path = ModelExportUtils.check_mar_already_exists(
            model_name, export_file_path, config.force, config.archive_format
        )

        # Step 2 : Copy all artifacts to temp directory
        artifact_files = {
            "model_file": model_file,
            "serialized_file": serialized_file,
            "handler": handler,
            "extra_files": extra_files,
            "requirements-file": requirements_file,
        }

        model_path = ModelExportUtils.copy_artifacts(model_name, **artifact_files)

        # Step 2 : Zip 'em all up
        ModelExportUtils.archive(
            export_file_path, model_name, model_path, manifest, config.archive_format
        )
        shutil.rmtree(model_path)
        logging.info(
            "Successfully exported model %s to file %s", model_name, export_file_path
        )
    except ModelArchiverError as e:
        logging.error(e)
        raise e


def generate_model_archive(config: ModelArchiverConfig):
    """
    Generate a model archive file
    :return:
    """

    logging.basicConfig(format="%(levelname)s - %(message)s")
    manifest = ModelExportUtils.generate_manifest_json(config)
    _package_model(config, manifest=manifest)

``` 

### Additional context

_No response_

I've tried setting up the environment following https://github.com/pytorch/serve/blob/master/CONTRIBUTING.md#install-torchserve-for-development and had some issues. Would you be able to help?

## Environment
I run on mac M1 pro with macOS Ventura. My typical setup is pyenv+poetry. I have no experience with node/java/gradle (anything but python essentially). The 2 scripts `python ./ts_scripts/install_dependencies.py --environment=dev` and `python ./ts_scripts/install_from_src.py --environment=dev` ran as far as I can tell.

## Issues

### 1. Installer does not create a python environment
I realised the script had installed packages to my "base" `pyenv` which isn't ideal. Would it be possible to leverage conda or any python env tool somewhere at the top of the [install_python_package()](https://github.com/pytorch/serve/blob/master/ts_scripts/install_dependencies.py#L53) function? Alternatively, it would be worth adding a disclaimer suggesting to create an environment before running the script.

### 2. Sanity check fails
When running the sanity check I get a grade error.

<details><summary>torchserve_sanity.py error</summary>

```bash
$ python torchserve_sanity.py
## Started frontend build and tests
## In directory: /Users/gauthiercastro/code/serve | Executing command: frontend/gradlew -p frontend clean build
Starting a Gradle Daemon (subsequent builds will be faster)

FAILURE: Build failed with an exception.

* What went wrong:
Could not open settings generic class cache for settings file '/Users/gauthiercastro/code/serve/frontend/settings.gradle' (/Users/gauthiercastro/.gradle/caches/7.6/scripts/9hv7ydusbul6bfoozqlmkymv6).
> BUG! exception in phase 'semantic analysis' in source unit '_BuildScript_' Unsupported class file major version 64

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2s
## Stopping TorchServe
## In directory: /Users/gauthiercastro/code/serve | Executing command: ['torchserve', '--stop']
## Successfully stopped TorchServe
## Deleting model_store_gen_dir: /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen

### Frontend Gradle Tests Failed !
```
</details>

### Regression tests fail
The regression_tests.py script seems to fail with similar complain about gradle but also `newman: requires node version >=16`
<details><summary>regression_tests.py error</summary>

```bash
$ python test/regression_tests.py
Current date and time : 2023-10-29 18:27:00
## In directory /Users/gauthiercastro/code/serve | Executing command pip install --force-reinstall .
Processing /Users/gauthiercastro/code/serve
  Preparing metadata (setup.py) ... done
Collecting Pillow (from torchserve==0.8.1b20231029)
  Using cached Pillow-10.1.0-cp38-cp38-macosx_11_0_arm64.whl.metadata (9.5 kB)
Collecting psutil (from torchserve==0.8.1b20231029)
  Using cached psutil-5.9.6-cp38-abi3-macosx_11_0_arm64.whl.metadata (21 kB)
Collecting packaging (from torchserve==0.8.1b20231029)
  Using cached packaging-23.2-py3-none-any.whl.metadata (3.2 kB)
Collecting wheel (from torchserve==0.8.1b20231029)
  Using cached wheel-0.41.2-py3-none-any.whl.metadata (2.2 kB)
Using cached packaging-23.2-py3-none-any.whl (53 kB)
Using cached Pillow-10.1.0-cp38-cp38-macosx_11_0_arm64.whl (3.3 MB)
Using cached psutil-5.9.6-cp38-abi3-macosx_11_0_arm64.whl (246 kB)
Using cached wheel-0.41.2-py3-none-any.whl (64 kB)
Building wheels for collected packages: torchserve
  Building wheel for torchserve (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [66 lines of output]
      -- torchserve building version: 0.8.1b20231029
      running bdist_wheel
      running build
      running build_py
      running build_frontend
      
      FAILURE: Build failed with an exception.
      
      * What went wrong:
      Could not open settings generic class cache for settings file '/Users/gauthiercastro/code/serve/frontend/settings.gradle' (/Users/gauthiercastro/.gradle/caches/7.6/scripts/9hv7ydusbul6bfoozqlmkymv6).
      > BUG! exception in phase 'semantic analysis' in source unit '_BuildScript_' Unsupported class file major version 64
      
      * Try:
      > Run with --stacktrace option to get the stack trace.
      > Run with --info or --debug option to get more log output.
      > Run with --scan to get full insights.
      
      * Get more help at https://help.gradle.org
      
      BUILD FAILED in 305ms
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/Users/gauthiercastro/code/serve/setup.py", line 163, in <module>
          setup(
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/__init__.py", line 103, in setup
          return distutils.core.setup(**attrs)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 185, in setup
          return run_commands(dist)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
          dist.run_commands()
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 969, in run_commands
          self.run_command(cmd)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/dist.py", line 989, in run_command
          super().run_command(command)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/wheel/bdist_wheel.py", line 364, in run
          self.run_command("build")
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
          self.distribution.run_command(command)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/dist.py", line 989, in run_command
          super().run_command(command)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/command/build.py", line 131, in run
          self.run_command(cmd_name)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
          self.distribution.run_command(command)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/dist.py", line 989, in run_command
          super().run_command(command)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "/Users/gauthiercastro/code/serve/setup.py", line 110, in run
          self.run_command("build_frontend")
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
          self.distribution.run_command(command)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/dist.py", line 989, in run_command
          super().run_command(command)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "/Users/gauthiercastro/code/serve/setup.py", line 97, in run
          subprocess.check_call(build_frontend_command[platform.system()], shell=True)
        File "/Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/subprocess.py", line 364, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command 'frontend/gradlew -p frontend clean assemble' returned non-zero exit status 1.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for torchserve
  Running setup.py clean for torchserve
Failed to build torchserve
ERROR: Could not build wheels for torchserve, which is required to install pyproject.toml-based projects
## In directory /Users/gauthiercastro/code/serve | Executing command pip install --force-reinstall model-archiver/.
Processing ./model-archiver
  Preparing metadata (setup.py) ... done
Collecting enum-compat (from torch-model-archiver==0.8.1b20231029)
  Using cached enum_compat-0.0.3-py3-none-any.whl (1.3 kB)
Building wheels for collected packages: torch-model-archiver
  Building wheel for torch-model-archiver (setup.py) ... done
  Created wheel for torch-model-archiver: filename=torch_model_archiver-0.8.1b20231029-py3-none-any.whl size=14829 sha256=d2bb68ad75f9ccb42c305611d87ba4a9a630b561cc77b93586cd165dbe4aa44b
  Stored in directory: /Users/gauthiercastro/Library/Caches/pip/wheels/22/b3/19/4406635d35836ce09ce29168013a196d6890fe544d303cf931
Successfully built torch-model-archiver
Installing collected packages: enum-compat, torch-model-archiver
  Attempting uninstall: enum-compat
    Found existing installation: enum-compat 0.0.3
    Uninstalling enum-compat-0.0.3:
      Successfully uninstalled enum-compat-0.0.3
  Attempting uninstall: torch-model-archiver
    Found existing installation: torch-model-archiver 0.8.1b20231029
    Uninstalling torch-model-archiver-0.8.1b20231029:
      Successfully uninstalled torch-model-archiver-0.8.1b20231029
Successfully installed enum-compat-0.0.3 torch-model-archiver-0.8.1b20231029
## In directory /Users/gauthiercastro/code/serve | Executing command pip install --force-reinstall workflow-archiver/.
Processing ./workflow-archiver
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: torch-workflow-archiver
  Building wheel for torch-workflow-archiver (setup.py) ... done
  Created wheel for torch-workflow-archiver: filename=torch_workflow_archiver-0.2.9b20231029-py3-none-any.whl size=12724 sha256=8e273d8695bc21dce5fde4d4a8851f29519f0b277af7bd7a63fb6b5918fcec52
  Stored in directory: /Users/gauthiercastro/Library/Caches/pip/wheels/09/54/e4/1592874f0b2465df1d159e0bdc11ea744780755b7c06f817e8
Successfully built torch-workflow-archiver
Installing collected packages: torch-workflow-archiver
  Attempting uninstall: torch-workflow-archiver
    Found existing installation: torch-workflow-archiver 0.2.9b20231029
    Uninstalling torch-workflow-archiver-0.2.9b20231029:
      Successfully uninstalled torch-workflow-archiver-0.2.9b20231029
Successfully installed torch-workflow-archiver-0.2.9b20231029
## Starting generate_mars, mar_config:/Users/gauthiercastro/code/serve/ts_scripts/../ts_scripts/mar_config.json, model_store_dir:/Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name fastrcnn --version 1.0 --model-file examples/object_detector/fast-rcnn/model.py --serialized-file /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/fasterrcnn_resnet50_fpn_coco-258fb6c6.pth --handler object_detector --extra-files examples/object_detector/index_to_name.json --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## fastrcnn.mar is generated.

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name alexnet --version 1.0 --model-file examples/image_classifier/alexnet/model.py --serialized-file /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/alexnet-owt-7be5be79.pth --handler image_classifier --extra-files examples/image_classifier/index_to_name.json --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## alexnet.mar is generated.

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name densenet161 --version 1.0 --model-file examples/image_classifier/densenet_161/model.py --serialized-file /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/densenet161-8d451a50.pth --handler image_classifier --extra-files examples/image_classifier/index_to_name.json --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## densenet161.mar is generated.

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name mnist --version 1.0 --model-file examples/image_classifier/mnist/mnist.py --serialized-file examples/image_classifier/mnist/mnist_cnn.pt --handler examples/image_classifier/mnist/mnist_handler.py --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## mnist.mar is generated.

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name resnet-152-batch --version 1.0 --model-file examples/image_classifier/resnet_152_batch/model.py --serialized-file /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/resnet152-394f9c45.pth --handler image_classifier --extra-files examples/image_classifier/index_to_name.json --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## resnet-152-batch.mar is generated.

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name resnet-18 --version 1.0 --model-file examples/image_classifier/resnet_18/model.py --serialized-file /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/resnet18-f37072fd.pth --handler image_classifier --extra-files examples/image_classifier/index_to_name.json --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## resnet-18.mar is generated.

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name squeezenet1_1 --version 1.0 --model-file examples/image_classifier/squeezenet/model.py --serialized-file /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/squeezenet1_1-b8a52dc0.pth --handler image_classifier --extra-files examples/image_classifier/index_to_name.json --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## squeezenet1_1.mar is generated.

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name vgg16 --version 1.0 --model-file examples/image_classifier/vgg_16/model.py --serialized-file /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/vgg16-397923af.pth --handler examples/image_classifier/vgg_16/vgg_handler.py --extra-files examples/image_classifier/index_to_name.json --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## vgg16.mar is generated.

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name deeplabv3_resnet_101_eager --version 1.0 --model-file examples/image_segmenter/deeplabv3/model.py --serialized-file /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/deeplabv3_resnet101_coco-586e9e4e.pth --handler image_segmenter --extra-files examples/image_segmenter/deeplabv3/deeplabv3.py,examples/image_segmenter/deeplabv3/intermediate_layer_getter.py,examples/image_segmenter/deeplabv3/fcn.py --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## deeplabv3_resnet_101_eager.mar is generated.

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name fcn_resnet_101 --version 1.0 --model-file examples/image_segmenter/fcn/model.py --serialized-file /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/fcn_resnet101_coco-7ecb50ca.pth --handler image_segmenter --extra-files examples/image_segmenter/fcn/fcn.py,examples/image_segmenter/fcn/intermediate_layer_getter.py --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## fcn_resnet_101.mar is generated.

## In directory: /Users/gauthiercastro/code/serve | Executing command: torch-model-archiver --model-name maskrcnn --version 1.0 --model-file examples/object_detector/maskrcnn/model.py --serialized-file /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/maskrcnn_resnet50_fpn_coco-bf2d0c1e.pth --handler object_detector --extra-files examples/object_detector/index_to_name.json --archive-format zip-store --export-path /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen --force

## maskrcnn.mar is generated.

## Starting gen_mar: model_store

## Create symlink for mar files

## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/squeezenet1_1.mar, model_store/squeezenet1_1.mar successfully.
## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/fastrcnn.mar, model_store/fastrcnn.mar successfully.
## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/resnet-152-batch.mar, model_store/resnet-152-batch.mar successfully.
## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/densenet161.mar, model_store/densenet161.mar successfully.
## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/maskrcnn.mar, model_store/maskrcnn.mar successfully.
## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/mnist.mar, model_store/mnist.mar successfully.
## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/resnet-18.mar, model_store/resnet-18.mar successfully.
## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/fcn_resnet_101.mar, model_store/fcn_resnet_101.mar successfully.
## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/deeplabv3_resnet_101_eager.mar, model_store/deeplabv3_resnet_101_eager.mar successfully.
## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/alexnet.mar, model_store/alexnet.mar successfully.
## Symlink /Users/gauthiercastro/code/serve/ts_scripts/../model_store_gen/vgg16.mar, model_store/vgg16.mar successfully.
## Starting TorchServe
## Console logs redirected to file: ts_console.log
## In directory: /Users/gauthiercastro/code/serve/test | Executing command: torchserve --start --model-store=model_store --ncs
## Successfully started TorchServe
newman: required node version >=16
## Stopping TorchServe
## In directory: /Users/gauthiercastro/code/serve/test | Executing command: ['torchserve', '--stop']
## Successfully stopped TorchServe
Traceback (most recent call last):
  File "test/regression_tests.py", line 28, in <module>
    test_api(
  File "/Users/gauthiercastro/code/serve/ts_scripts/api_utils.py", line 432, in test_api
    exit_code = switcher[collection]()
  File "/Users/gauthiercastro/code/serve/ts_scripts/api_utils.py", line 359, in trigger_all
    exit_code1 = trigger_management_tests()
  File "/Users/gauthiercastro/code/serve/ts_scripts/api_utils.py", line 126, in trigger_management_tests
    move_logs(TS_CONSOLE_LOG_FILE, ARTIFACTS_MANAGEMENT_DIR)
  File "/Users/gauthiercastro/code/serve/ts_scripts/api_utils.py", line 113, in move_logs
    os.rename(log_file, os.path.join(logs_dir, log_file))  # mv file logs/
FileNotFoundError: [Errno 2] No such file or directory: 'ts_console.log' -> 'logs/ts_console.log'
```

</details> 


##¢ Pytest fails to find python module `inference_pb2`

<details><summary>missing `inference_pb2`</summary>

```bash
pytest test/pytest 
======================================================================= test session starts ========================================================================
platform darwin -- Python 3.8.14, pytest-7.3.1, pluggy-1.0.0
rootdir: /Users/gauthiercastro/code/serve
plugins: mock-3.10.0, cov-4.1.0
collected 106 items / 3 errors                                                                                                                                     

============================================================================== ERRORS ==============================================================================
_____________________________________________________ ERROR collecting test/pytest/test_gRPC_inference_api.py ______________________________________________________
ImportError while importing test module '/Users/gauthiercastro/code/serve/test/pytest/test_gRPC_inference_api.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../.pyenv/versions/3.8.14/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test/pytest/test_gRPC_inference_api.py:5: in <module>
    import inference_pb2
E   ModuleNotFoundError: No module named 'inference_pb2'
____________________________________________________ ERROR collecting test/pytest/test_gRPC_management_apis.py _____________________________________________________
ImportError while importing test module '/Users/gauthiercastro/code/serve/test/pytest/test_gRPC_management_apis.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../.pyenv/versions/3.8.14/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test/pytest/test_gRPC_management_apis.py:4: in <module>
    import test_gRPC_utils
test/pytest/test_gRPC_utils.py:2: in <module>
    import inference_pb2_grpc
E   ModuleNotFoundError: No module named 'inference_pb2_grpc'
_________________________________________________________ ERROR collecting test/pytest/test_gRPC_utils.py __________________________________________________________
ImportError while importing test module '/Users/gauthiercastro/code/serve/test/pytest/test_gRPC_utils.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../.pyenv/versions/3.8.14/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test/pytest/test_gRPC_utils.py:2: in <module>
    import inference_pb2_grpc
E   ModuleNotFoundError: No module named 'inference_pb2_grpc'
========================================================================= warnings summary =========================================================================
test/pytest/test_torch_compile.py:11
  /Users/gauthiercastro/code/serve/test/pytest/test_torch_compile.py:11: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    from pkg_resources import packaging

../../.pyenv/versions/3.8.14/lib/python3.8/site-packages/pkg_resources/__init__.py:2871
  /Users/gauthiercastro/.pyenv/versions/3.8.14/lib/python3.8/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('mpl_toolkits')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================================================================== short test summary info ======================================================================
ERROR test/pytest/test_gRPC_inference_api.py
ERROR test/pytest/test_gRPC_management_apis.py
ERROR test/pytest/test_gRPC_utils.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================== 2 warnings, 3 errors in 1.69s ===================================================================
```

</details>

--- a/model-archiver/model_archiver/__init__.py
+++ b/model-archiver/model_archiver/__init__.py
@@ -1,5 +1,3 @@
-
-
 """
 This module does the following:
 Exports the model folder to generate a Model Archive file out of it in .mar format
@@ -7,3 +5,6 @@ Exports the model folder to generate a Model Archive file out of it in .mar form
 from . import version
 
 __version__ = version.__version__
+
+from .model_archiver import ModelArchiver
+from .model_archiver_config import ModelArchiverConfig

--- a/model-archiver/model_archiver/arg_parser.py
+++ b/model-archiver/model_archiver/arg_parser.py
@@ -6,6 +6,7 @@ import argparse
 import os
 
 from .manifest_components.manifest import RuntimeType
+from .model_archiver_config import ModelArchiverConfig
 
 
 # noinspection PyTypeChecker
@@ -154,4 +155,4 @@ class ArgParser(object):
             help="Path to a yaml file containing model configuration eg. batch_size.",
         )
 
-        return parser_export
+        return ModelArchiverConfig.from_args(parser_export.parse_args())

--- a/model-archiver/model_archiver/model_archiver.py
+++ b/model-archiver/model_archiver/model_archiver.py
@@ -1,5 +1,5 @@
 """
-Command line interface to export model files to be used for inference by MXNet Model Server
+Helper class to generate a model archive file
 """
 
 from model_archiver.model_archiver_config import ModelArchiverConfig

--- a/model-archiver/model_archiver/model_archiver.py
+++ b/model-archiver/model_archiver/model_archiver.py
@@ -1,0 +1,17 @@
+"""
+Command line interface to export model files to be used for inference by MXNet Model Server
+"""
+
+from model_archiver.model_archiver_config import ModelArchiverConfig
+from model_archiver.model_packaging import generate_model_archive
+
+
+class ModelArchiver:
+    @staticmethod
+    def generate_model_archive(config: ModelArchiverConfig) -> None:
+        """
+        Generate a model archive file
+        :param config: Model Archiver Config object
+        :return:
+        """
+        generate_model_archive(config)

--- a/model-archiver/model_archiver/model_archiver_config.py
+++ b/model-archiver/model_archiver/model_archiver_config.py
@@ -35,12 +35,7 @@ class ModelArchiverConfig:
         self.config_file = config_file
 
     @staticmethod
-    def from_args(args: Optional[Namespace] = None) -> "ModelArchiverConfig":
-        if args is None:
-            from model_archiver.arg_parser import ArgParser
-
-            args = ArgParser.export_model_args_parser().parse_args()
-
+    def from_args(args: Namespace) -> "ModelArchiverConfig":
         config = ModelArchiverConfig(
             model_name=args.model_name,
             handler=args.handler,

--- a/model-archiver/model_archiver/model_archiver_config.py
+++ b/model-archiver/model_archiver/model_archiver_config.py
@@ -1,0 +1,58 @@
+import os
+from argparse import Namespace
+from typing import Literal, Optional
+
+from model_archiver.manifest_components.manifest import RuntimeType
+
+
+class ModelArchiverConfig:
+    def __init__(
+        self,
+        model_name: str,
+        handler: str,
+        version: str,
+        serialized_file: Optional[str] = None,
+        model_file: Optional[str] = None,
+        extra_files: Optional[str] = None,
+        runtime: str = RuntimeType.PYTHON.value,
+        export_path: str = os.getcwd(),
+        archive_format: Literal["default", "tgz", "no-archive"] = "default",
+        force: bool = False,
+        requirements_file: Optional[str] = None,
+        config_file: Optional[str] = None,
+    ):
+        self.model_name = model_name
+        self.handler = handler
+        self.version = version
+        self.serialized_file = serialized_file
+        self.model_file = model_file
+        self.extra_files = extra_files
+        self.runtime = runtime
+        self.export_path = export_path
+        self.archive_format = archive_format
+        self.force = force
+        self.requirements_file = requirements_file
+        self.config_file = config_file
+
+    @staticmethod
+    def from_args(args: Optional[Namespace] = None) -> "ModelArchiverConfig":
+        if args is None:
+            from model_archiver.arg_parser import ArgParser
+
+            args = ArgParser.export_model_args_parser().parse_args()
+
+        config = ModelArchiverConfig(
+            model_name=args.model_name,
+            handler=args.handler,
+            version=args.version,
+            serialized_file=args.serialized_file,
+            model_file=args.model_file,
+            extra_files=args.extra_files,
+            runtime=args.runtime,
+            export_path=args.export_path,
+            archive_format=args.archive_format,
+            force=args.force,
+            requirements_file=args.requirements_file,
+            config_file=args.config_file,
+        )
+        return config

--- a/model-archiver/model_archiver/model_archiver_config.py
+++ b/model-archiver/model_archiver/model_archiver_config.py
@@ -1,38 +1,25 @@
 import os
 from argparse import Namespace
+from dataclasses import dataclass
 from typing import Literal, Optional
 
 from model_archiver.manifest_components.manifest import RuntimeType
 
 
+@dataclass
 class ModelArchiverConfig:
-    def __init__(
-        self,
-        model_name: str,
-        handler: str,
-        version: str,
-        serialized_file: Optional[str] = None,
-        model_file: Optional[str] = None,
-        extra_files: Optional[str] = None,
-        runtime: str = RuntimeType.PYTHON.value,
-        export_path: str = os.getcwd(),
-        archive_format: Literal["default", "tgz", "no-archive"] = "default",
-        force: bool = False,
-        requirements_file: Optional[str] = None,
-        config_file: Optional[str] = None,
-    ):
-        self.model_name = model_name
-        self.handler = handler
-        self.version = version
-        self.serialized_file = serialized_file
-        self.model_file = model_file
-        self.extra_files = extra_files
-        self.runtime = runtime
-        self.export_path = export_path
-        self.archive_format = archive_format
-        self.force = force
-        self.requirements_file = requirements_file
-        self.config_file = config_file
+    model_name: str
+    handler: str
+    version: str
+    serialized_file: Optional[str] = None
+    model_file: Optional[str] = None
+    extra_files: Optional[str] = None
+    runtime: str = RuntimeType.PYTHON.value
+    export_path: str = os.getcwd()
+    archive_format: Literal["default", "tgz", "no-archive"] = "default"
+    force: bool = False
+    requirements_file: Optional[str] = None
+    config_file: Optional[str] = None
 
     @staticmethod
     def from_args(args: Namespace) -> "ModelArchiverConfig":

--- a/model-archiver/model_archiver/model_archiver_config.py
+++ b/model-archiver/model_archiver/model_archiver_config.py
@@ -1,6 +1,6 @@
 import os
 from argparse import Namespace
-from dataclasses import dataclass
+from dataclasses import dataclass, fields
 from typing import Literal, Optional
 
 from model_archiver.manifest_components.manifest import RuntimeType
@@ -21,20 +21,8 @@ class ModelArchiverConfig:
     requirements_file: Optional[str] = None
     config_file: Optional[str] = None
 
-    @staticmethod
-    def from_args(args: Namespace) -> "ModelArchiverConfig":
-        config = ModelArchiverConfig(
-            model_name=args.model_name,
-            handler=args.handler,
-            version=args.version,
-            serialized_file=args.serialized_file,
-            model_file=args.model_file,
-            extra_files=args.extra_files,
-            runtime=args.runtime,
-            export_path=args.export_path,
-            archive_format=args.archive_format,
-            force=args.force,
-            requirements_file=args.requirements_file,
-            config_file=args.config_file,
-        )
+    @classmethod
+    def from_args(cls, args: Namespace) -> "ModelArchiverConfig":
+        params = {field.name: getattr(args, field.name) for field in fields(cls)}
+        config = cls(**params)
         return config

--- a/model-archiver/model_archiver/model_packaging.py
+++ b/model-archiver/model_archiver/model_packaging.py
@@ -65,8 +65,7 @@ def generate_model_archive(config: Optional[ModelArchiverConfig] = None):
 
     logging.basicConfig(format="%(levelname)s - %(message)s")
     if config is None:
-        args = ArgParser.export_model_args_parser().parse_args()
-        config = ModelArchiverConfig.from_args(args)
+        config = ArgParser.export_model_args_parser()
     manifest = ModelExportUtils.generate_manifest_json(config)
     package_model(config, manifest=manifest)
 

--- a/model-archiver/model_archiver/model_packaging.py
+++ b/model-archiver/model_archiver/model_packaging.py
@@ -4,31 +4,32 @@ Command line interface to export model files to be used for inference by MXNet M
 
 import logging
 import shutil
-import sys
+from typing import Optional
 
 from model_archiver.arg_parser import ArgParser
+from model_archiver.model_archiver_config import ModelArchiverConfig
 from model_archiver.model_archiver_error import ModelArchiverError
 from model_archiver.model_packaging_utils import ModelExportUtils
 
 
-def package_model(args, manifest):
+def package_model(config: ModelArchiverConfig, manifest: str):
     """
     Internal helper for the exporting model command line interface.
     """
-    model_file = args.model_file
-    serialized_file = args.serialized_file
-    model_name = args.model_name
-    handler = args.handler
-    extra_files = args.extra_files
-    export_file_path = args.export_path
-    requirements_file = args.requirements_file
-    config_file = args.config_file
+    model_file = config.model_file
+    serialized_file = config.serialized_file
+    model_name = config.model_name
+    handler = config.handler
+    extra_files = config.extra_files
+    export_file_path = config.export_path
+    requirements_file = config.requirements_file
+    config_file = config.config_file
 
     try:
         ModelExportUtils.validate_inputs(model_name, export_file_path)
         # Step 1 : Check if .mar already exists with the given model name
         export_file_path = ModelExportUtils.check_mar_already_exists(
-            model_name, export_file_path, args.force, args.archive_format
+            model_name, export_file_path, config.force, config.archive_format
         )
 
         # Step 2 : Copy all artifacts to temp directory
@@ -45,7 +46,7 @@ def package_model(args, manifest):
 
         # Step 2 : Zip 'em all up
         ModelExportUtils.archive(
-            export_file_path, model_name, model_path, manifest, args.archive_format
+            export_file_path, model_name, model_path, manifest, config.archive_format
         )
         shutil.rmtree(model_path)
         logging.info(
@@ -53,19 +54,21 @@ def package_model(args, manifest):
         )
     except ModelArchiverError as e:
         logging.error(e)
-        sys.exit(1)
+        raise e
 
 
-def generate_model_archive():
+def generate_model_archive(config: Optional[ModelArchiverConfig] = None):
     """
     Generate a model archive file
     :return:
     """
 
     logging.basicConfig(format="%(levelname)s - %(message)s")
-    args = ArgParser.export_model_args_parser().parse_args()
-    manifest = ModelExportUtils.generate_manifest_json(args)
-    package_model(args, manifest=manifest)
+    if config is None:
+        args = ArgParser.export_model_args_parser().parse_args()
+        config = ModelArchiverConfig.from_args(args)
+    manifest = ModelExportUtils.generate_manifest_json(config)
+    package_model(config, manifest=manifest)
 
 
 if __name__ == "__main__":

--- a/model-archiver/model_archiver/model_packaging_utils.py
+++ b/model-archiver/model_archiver/model_packaging_utils.py
@@ -15,6 +15,7 @@ from pathlib import Path
 
 from .manifest_components.manifest import Manifest
 from .manifest_components.model import Model
+from .model_archiver_config import ModelArchiverConfig
 from .model_archiver_error import ModelArchiverError
 
 archiving_options = {
@@ -107,29 +108,29 @@ class ModelExportUtils(object):
             )
 
     @staticmethod
-    def generate_model(modelargs):
+    def generate_model(modelcfg: ModelArchiverConfig):
         model = Model(
-            model_name=modelargs.model_name,
-            serialized_file=modelargs.serialized_file,
-            model_file=modelargs.model_file,
-            handler=modelargs.handler,
-            model_version=modelargs.version,
-            requirements_file=modelargs.requirements_file,
-            config_file=modelargs.config_file,
+            model_name=modelcfg.model_name,
+            serialized_file=modelcfg.serialized_file,
+            model_file=modelcfg.model_file,
+            handler=modelcfg.handler,
+            model_version=modelcfg.version,
+            requirements_file=modelcfg.requirements_file,
+            config_file=modelcfg.config_file,
         )
         return model
 
     @staticmethod
-    def generate_manifest_json(args):
+    def generate_manifest_json(config: ModelArchiverConfig) -> str:
         """
         Function to generate manifest as a json string from the inputs provided by the user in the command line
         :param args:
         :return:
         """
 
-        model = ModelExportUtils.generate_model(args)
+        model = ModelExportUtils.generate_model(config)
 
-        manifest = Manifest(runtime=args.runtime, model=model)
+        manifest = Manifest(runtime=config.runtime, model=model)
 
         return str(manifest)
 

--- a/model-archiver/model_archiver/tests/integ_tests/test_integration_model_archiver.py
+++ b/model-archiver/model_archiver/tests/integ_tests/test_integration_model_archiver.py
@@ -41,10 +41,10 @@ def delete_file_path(path):
 
 
 def run_test(test, config, mocker):
-    m = mocker.patch(
+    mocker.patch(
         "model_archiver.model_packaging.ArgParser.export_model_args_parser",
+        return_value=config,
     )
-    m.return_value = config
     from model_archiver.model_packaging import generate_model_archive
 
     it = test.get("iterations", 1)

--- a/model-archiver/model_archiver/tests/integ_tests/test_integration_model_archiver.py
+++ b/model-archiver/model_archiver/tests/integ_tests/test_integration_model_archiver.py
@@ -45,7 +45,6 @@ def run_test(test, args, mocker):
         "model_archiver.model_packaging.ArgParser.export_model_args_parser",
     )
     m.return_value.parse_args.return_value = args
-    mocker.patch("sys.exit", side_effect=Exception())
     from model_archiver.model_packaging import generate_model_archive
 
     it = test.get("iterations", 1)

--- a/model-archiver/model_archiver/tests/integ_tests/test_integration_model_archiver.py
+++ b/model-archiver/model_archiver/tests/integ_tests/test_integration_model_archiver.py
@@ -40,11 +40,11 @@ def delete_file_path(path):
         pass
 
 
-def run_test(test, args, mocker):
+def run_test(test, config, mocker):
     m = mocker.patch(
         "model_archiver.model_packaging.ArgParser.export_model_args_parser",
     )
-    m.return_value.parse_args.return_value = args
+    m.return_value = config
     from model_archiver.model_packaging import generate_model_archive
 
     it = test.get("iterations", 1)
@@ -178,7 +178,9 @@ def build_namespace(test):
 
     args = Namespace(**{k.replace("-", "_"): test[k] for k in keys})
 
-    return args
+    config = model_archiver.ModelArchiverConfig.from_args(args)
+
+    return config
 
 
 def make_paths_absolute(test, keys):

--- a/model-archiver/model_archiver/tests/unit_tests/test_model_archiver.py
+++ b/model-archiver/model_archiver/tests/unit_tests/test_model_archiver.py
@@ -1,0 +1,77 @@
+from collections import namedtuple
+
+import pytest
+from model_archiver import ModelArchiver, ModelArchiverConfig
+from model_archiver.manifest_components.manifest import RuntimeType
+
+
+# noinspection PyClassHasNoInit
+class TestModelArchiver:
+    model_name = "my-model"
+    model_file = "my-model/"
+    serialized_file = "my-model/"
+    handler = "a.py::my-awesome-func"
+    export_path = "/Users/dummyUser/"
+    version = "1.0"
+    requirements_file = "requirements.txt"
+    config_file = None
+    source_vocab = None
+
+    config = ModelArchiverConfig(
+        model_name=model_name,
+        handler=handler,
+        runtime=RuntimeType.PYTHON.value,
+        model_file=model_file,
+        serialized_file=serialized_file,
+        extra_files=None,
+        export_path=export_path,
+        force=False,
+        archive_format="default",
+        version=version,
+        requirements_file=requirements_file,
+        config_file=None,
+    )
+
+    @pytest.fixture()
+    def patches(self, mocker):
+        Patches = namedtuple("Patches", ["arg_parse", "export_utils", "export_method"])
+        patches = Patches(
+            mocker.patch("model_archiver.arg_parser.ArgParser"),
+            mocker.patch("model_archiver.model_packaging.ModelExportUtils"),
+            mocker.patch("model_archiver.model_packaging.package_model"),
+        )
+        mocker.patch("shutil.rmtree")
+
+        return patches
+
+    def test_gen_model_archive(self, patches):
+        ModelArchiver.generate_model_archive(self.config)
+        patches.export_method.assert_called()
+
+    def test_model_archiver_config_from_args(self, patches):
+        class Namespace:
+            def __init__(self, **kwargs):
+                self.__dict__.update(kwargs)
+
+            def update(self, **kwargs):
+                self.__dict__.update(kwargs)
+
+        patches.arg_parse.export_model_args_parser.parse_args.return_value = Namespace(
+            model_name=self.model_name,
+            handler=self.handler,
+            runtime=RuntimeType.PYTHON.value,
+            model_file=self.model_file,
+            serialized_file=self.serialized_file,
+            extra_files=None,
+            export_path=self.export_path,
+            force=False,
+            archive_format="default",
+            convert=False,
+            version=self.version,
+            source_vocab=self.source_vocab,
+            requirements_file=self.requirements_file,
+            config_file=None,
+        )
+        config = ModelArchiverConfig.from_args(None)
+
+        config == self.config

--- a/model-archiver/model_archiver/tests/unit_tests/test_model_archiver.py
+++ b/model-archiver/model_archiver/tests/unit_tests/test_model_archiver.py
@@ -65,4 +65,4 @@ class TestModelArchiver:
         )
         config = ModelArchiverConfig.from_args(args)
 
-        config == self.config
+        assert config == self.config

--- a/model-archiver/model_archiver/tests/unit_tests/test_model_archiver.py
+++ b/model-archiver/model_archiver/tests/unit_tests/test_model_archiver.py
@@ -1,3 +1,4 @@
+from argparse import Namespace
 from collections import namedtuple
 
 import pytest
@@ -48,15 +49,8 @@ class TestModelArchiver:
         ModelArchiver.generate_model_archive(self.config)
         patches.export_method.assert_called()
 
-    def test_model_archiver_config_from_args(self, patches):
-        class Namespace:
-            def __init__(self, **kwargs):
-                self.__dict__.update(kwargs)
-
-            def update(self, **kwargs):
-                self.__dict__.update(kwargs)
-
-        patches.arg_parse.export_model_args_parser.parse_args.return_value = Namespace(
+    def test_model_archiver_config_from_args(self):
+        args = Namespace(
             model_name=self.model_name,
             handler=self.handler,
             runtime=RuntimeType.PYTHON.value,
@@ -72,6 +66,6 @@ class TestModelArchiver:
             requirements_file=self.requirements_file,
             config_file=None,
         )
-        config = ModelArchiverConfig.from_args(None)
+        config = ModelArchiverConfig.from_args(args)
 
         config == self.config

--- a/model-archiver/model_archiver/tests/unit_tests/test_model_archiver.py
+++ b/model-archiver/model_archiver/tests/unit_tests/test_model_archiver.py
@@ -16,7 +16,6 @@ class TestModelArchiver:
     version = "1.0"
     requirements_file = "requirements.txt"
     config_file = None
-    source_vocab = None
 
     config = ModelArchiverConfig(
         model_name=model_name,
@@ -60,9 +59,7 @@ class TestModelArchiver:
             export_path=self.export_path,
             force=False,
             archive_format="default",
-            convert=False,
             version=self.version,
-            source_vocab=self.source_vocab,
             requirements_file=self.requirements_file,
             config_file=None,
         )

--- a/model-archiver/model_archiver/tests/unit_tests/test_model_packaging.py
+++ b/model-archiver/model_archiver/tests/unit_tests/test_model_packaging.py
@@ -1,6 +1,7 @@
 from collections import namedtuple
 
 import pytest
+from model_archiver import ModelArchiverConfig
 from model_archiver.manifest_components.manifest import RuntimeType
 from model_archiver.model_packaging import generate_model_archive, package_model
 from model_archiver.model_packaging_utils import ModelExportUtils
@@ -8,13 +9,6 @@ from model_archiver.model_packaging_utils import ModelExportUtils
 
 # noinspection PyClassHasNoInit
 class TestModelPackaging:
-    class Namespace:
-        def __init__(self, **kwargs):
-            self.__dict__.update(kwargs)
-
-        def update(self, **kwargs):
-            self.__dict__.update(kwargs)
-
     model_name = "my-model"
     model_file = "my-model/"
     serialized_file = "my-model/"
@@ -23,9 +17,8 @@ class TestModelPackaging:
     version = "1.0"
     requirements_file = "requirements.txt"
     config_file = None
-    source_vocab = None
 
-    args = Namespace(
+    config = ModelArchiverConfig(
         model_name=model_name,
         handler=handler,
         runtime=RuntimeType.PYTHON.value,
@@ -35,9 +28,7 @@ class TestModelPackaging:
         export_path=export_path,
         force=False,
         archive_format="default",
-        convert=False,
         version=version,
-        source_vocab=source_vocab,
         requirements_file=requirements_file,
         config_file=None,
     )
@@ -55,7 +46,7 @@ class TestModelPackaging:
         return patches
 
     def test_gen_model_archive(self, patches):
-        patches.arg_parse.export_model_args_parser.parse_args.return_value = self.args
+        patches.arg_parse.export_model_args_parser.parse_args.return_value = self.config
         generate_model_archive()
         patches.export_method.assert_called()
 
@@ -67,12 +58,12 @@ class TestModelPackaging:
         )
         patches.export_utils.zip.return_value = None
 
-        package_model(self.args, ModelExportUtils.generate_manifest_json(self.args))
+        package_model(self.config, ModelExportUtils.generate_manifest_json(self.config))
         patches.export_utils.validate_inputs.assert_called()
         patches.export_utils.archive.assert_called()
 
     def test_export_model_method_tar(self, patches):
-        self.args.update(archive_format="tar")
+        self.config.archive_format = "tar"
         patches.export_utils.check_mar_already_exists.return_value = "/Users/dummyUser/"
         patches.export_utils.check_custom_model_types.return_value = (
             "/Users/dummyUser",
@@ -80,12 +71,12 @@ class TestModelPackaging:
         )
         patches.export_utils.zip.return_value = None
 
-        package_model(self.args, ModelExportUtils.generate_manifest_json(self.args))
+        package_model(self.config, ModelExportUtils.generate_manifest_json(self.config))
         patches.export_utils.validate_inputs.assert_called()
         patches.export_utils.archive.assert_called()
 
     def test_export_model_method_noarchive(self, patches):
-        self.args.update(archive_format="no-archive")
+        self.config.archive_format = "no-archive"
         patches.export_utils.check_mar_already_exists.return_value = "/Users/dummyUser/"
         patches.export_utils.check_custom_model_types.return_value = (
             "/Users/dummyUser",
@@ -93,6 +84,6 @@ class TestModelPackaging:
         )
         patches.export_utils.zip.return_value = None
 
-        package_model(self.args, ModelExportUtils.generate_manifest_json(self.args))
+        package_model(self.config, ModelExportUtils.generate_manifest_json(self.config))
         patches.export_utils.validate_inputs.assert_called()
         patches.export_utils.archive.assert_called()

--- a/model-archiver/model_archiver/tests/unit_tests/test_model_packaging.py
+++ b/model-archiver/model_archiver/tests/unit_tests/test_model_packaging.py
@@ -63,13 +63,12 @@ class TestModelPackaging:
         patches.export_utils.archive.assert_called()
 
     def test_export_model_method_tar(self, patches):
-        self.config.archive_format = "tar"
+        self.config.archive_format = "tgz"
         patches.export_utils.check_mar_already_exists.return_value = "/Users/dummyUser/"
         patches.export_utils.check_custom_model_types.return_value = (
             "/Users/dummyUser",
             ["a.txt", "b.txt"],
         )
-        patches.export_utils.zip.return_value = None
 
         package_model(self.config, ModelExportUtils.generate_manifest_json(self.config))
         patches.export_utils.validate_inputs.assert_called()

--- a/model-archiver/model_archiver/tests/unit_tests/test_model_packaging_utils.py
+++ b/model-archiver/model_archiver/tests/unit_tests/test_model_packaging_utils.py
@@ -4,6 +4,7 @@ from collections import namedtuple
 from pathlib import Path
 
 import pytest
+from model_archiver import ModelArchiverConfig
 from model_archiver.manifest_components.manifest import RuntimeType
 from model_archiver.model_archiver_error import ModelArchiverError
 from model_archiver.model_packaging_utils import ModelExportUtils
@@ -142,10 +143,6 @@ class TestExportModelUtils:
 
     # noinspection PyClassHasNoInit
     class TestGenerateManifestProps:
-        class Namespace:
-            def __init__(self, **kwargs):
-                self.__dict__.update(kwargs)
-
         model_name = "my-model"
         handler = "a.py::my-awesome-func"
         serialized_file = "model.pt"
@@ -153,7 +150,7 @@ class TestExportModelUtils:
         version = "1.0"
         requirements_file = "requirements.txt"
 
-        args = Namespace(
+        config = ModelArchiverConfig(
             model_name=model_name,
             handler=handler,
             runtime=RuntimeType.PYTHON.value,
@@ -165,12 +162,12 @@ class TestExportModelUtils:
         )
 
         def test_model(self):
-            mod = ModelExportUtils.generate_model(self.args)
+            mod = ModelExportUtils.generate_model(self.config)
             assert mod.model_name == self.model_name
             assert mod.handler == self.handler
 
         def test_manifest_json(self):
-            manifest = ModelExportUtils.generate_manifest_json(self.args)
+            manifest = ModelExportUtils.generate_manifest_json(self.config)
             manifest_json = json.loads(manifest)
             assert manifest_json["runtime"] == RuntimeType.PYTHON.value
             assert "model" in manifest_json

--- a/test/pytest/test_auto_recover.py
+++ b/test/pytest/test_auto_recover.py
@@ -1,13 +1,13 @@
 import json
 import platform
 import shutil
-from argparse import Namespace
 from pathlib import Path
-from unittest.mock import MagicMock, patch
+from unittest.mock import patch
 
 import pytest
 import requests
 import test_utils
+from model_archiver import ModelArchiverConfig
 
 CURR_FILE_PATH = Path(__file__).parent
 REPO_ROOT_DIR = CURR_FILE_PATH.parent.parent
@@ -97,7 +97,7 @@ def create_mar_file(work_dir, model_archiver, model_name):
     handler_py_file = work_dir / "handler.py"
     handler_py_file.write_text(HANDLER_PY)
 
-    args = Namespace(
+    config = ModelArchiverConfig(
         model_name=model_name,
         version="1.0",
         serialized_file=None,
@@ -112,9 +112,7 @@ def create_mar_file(work_dir, model_archiver, model_name):
         config_file=None,
     )
 
-    mock = MagicMock()
-    mock.parse_args = MagicMock(return_value=args)
-    with patch("archiver.ArgParser.export_model_args_parser", return_value=mock):
+    with patch("archiver.ArgParser.export_model_args_parser", return_value=config):
         model_archiver.generate_model_archive()
 
         assert mar_file_path.exists()

--- a/test/pytest/test_continuous_batching.py
+++ b/test/pytest/test_continuous_batching.py
@@ -1,15 +1,15 @@
 import json
 import shutil
-from argparse import Namespace
 from pathlib import Path
 from queue import Empty
-from unittest.mock import MagicMock, patch
+from unittest.mock import patch
 from zipfile import ZIP_STORED, ZipFile
 
 import pytest
 import requests
 import test_utils
 import torch
+from model_archiver import ModelArchiverConfig
 from test_data.streaming.stream_handler import StreamingHandler
 
 from ts.torch_handler.unit_tests.test_utils.mock_context import MockContext
@@ -31,7 +31,7 @@ def work_dir(tmp_path_factory, model_name):
 def create_mar_file(work_dir, model_archiver, model_name):
     mar_file_path = Path(work_dir).joinpath(model_name + ".mar")
 
-    args = Namespace(
+    config = ModelArchiverConfig(
         model_name=model_name,
         version="1.0",
         model_file=CURR_FILE_PATH.joinpath(
@@ -52,9 +52,7 @@ def create_mar_file(work_dir, model_archiver, model_name):
         extra_files=None,
     )
 
-    mock = MagicMock()
-    mock.parse_args = MagicMock(return_value=args)
-    with patch("archiver.ArgParser.export_model_args_parser", return_value=mock):
+    with patch("archiver.ArgParser.export_model_args_parser", return_value=config):
         # Using ZIP_STORED instead of ZIP_DEFLATED reduces test runtime from 54 secs to 10 secs
         with patch(
             "model_archiver.model_packaging_utils.zipfile.ZipFile",

--- a/test/pytest/test_example_micro_batching.py
+++ b/test/pytest/test_example_micro_batching.py
@@ -2,15 +2,15 @@ import asyncio
 import json
 import random
 import shutil
-from argparse import Namespace
 from pathlib import Path
-from unittest.mock import MagicMock, patch
+from unittest.mock import patch
 from zipfile import ZIP_STORED, ZipFile
 
 import pytest
 import requests
 import test_utils
 import yaml
+from model_archiver import ModelArchiverConfig
 from torchvision.models.resnet import ResNet18_Weights
 
 from ts.torch_handler.unit_tests.test_utils.model_dir import download_model
@@ -113,7 +113,7 @@ def create_mar_file(work_dir, serialized_file, model_archiver, model_name, reque
 
     extra_files = [name_file]
 
-    args = Namespace(
+    config = ModelArchiverConfig(
         model_name=model_name,
         version="1.0",
         serialized_file=str(serialized_file),
@@ -132,9 +132,7 @@ def create_mar_file(work_dir, serialized_file, model_archiver, model_name, reque
         config_file=config_file,
     )
 
-    mock = MagicMock()
-    mock.parse_args = MagicMock(return_value=args)
-    with patch("archiver.ArgParser.export_model_args_parser", return_value=mock):
+    with patch("archiver.ArgParser.export_model_args_parser", return_value=config):
         # Using ZIP_STORED instead of ZIP_DEFLATED reduces test runtime from 54 secs to 10 secs
         with patch(
             "model_archiver.model_packaging_utils.zipfile.ZipFile",

--- a/test/pytest/test_example_near_real_time_video.py
+++ b/test/pytest/test_example_near_real_time_video.py
@@ -4,13 +4,13 @@ Unit test for near real-time video example
 import json
 import os
 import shutil
-from argparse import Namespace
 from pathlib import Path
 
 import pytest
 import requests
 import test_utils
 import torch
+from model_archiver import ModelArchiverConfig
 
 from ts.torch_handler.image_classifier import ImageClassifier
 from ts.torch_handler.unit_tests.test_utils.mock_context import MockContext
@@ -50,7 +50,7 @@ def create_mar_file(work_dir, session_mocker, model_archiver):
 
     mar_file_path = Path(work_dir).joinpath(model_name + ".mar")
 
-    args = Namespace(
+    config = ModelArchiverConfig(
         model_name=model_name,
         version="1.0",
         serialized_file=os.path.join(REPO_ROOT_DIR, MODEL_PTH_FILE),
@@ -67,10 +67,8 @@ def create_mar_file(work_dir, session_mocker, model_archiver):
         config_file=None,
     )
 
-    mock = session_mocker.MagicMock()
-    mock.parse_args = session_mocker.MagicMock(return_value=args)
     session_mocker.patch(
-        "archiver.ArgParser.export_model_args_parser", return_value=mock
+        "archiver.ArgParser.export_model_args_parser", return_value=config
     )
 
     # Using ZIP_STORED instead of ZIP_DEFLATED reduces test runtime from 54 secs to 10 secs

--- a/test/pytest/test_example_scriptable_tokenzier.py
+++ b/test/pytest/test_example_scriptable_tokenzier.py
@@ -11,6 +11,7 @@ import pytest
 import requests
 import test_utils
 import torch
+from model_archiver import ModelArchiverConfig
 from test_utils import REPO_ROOT
 
 from ts.torch_handler.unit_tests.test_utils.mock_context import MockContext
@@ -148,7 +149,7 @@ def create_mar_file(work_dir, session_mocker, jit_file_path, model_archiver):
 
     mar_file_path = os.path.join(work_dir, model_name + ".mar")
 
-    args = Namespace(
+    config = ModelArchiverConfig(
         model_name=model_name,
         version="1.0",
         serialized_file=jit_file_path,
@@ -163,10 +164,8 @@ def create_mar_file(work_dir, session_mocker, jit_file_path, model_archiver):
         config_file=None,
     )
 
-    mock = session_mocker.MagicMock()
-    mock.parse_args = session_mocker.MagicMock(return_value=args)
     session_mocker.patch(
-        "archiver.ArgParser.export_model_args_parser", return_value=mock
+        "archiver.ArgParser.export_model_args_parser", return_value=config
     )
 
     # Using ZIP_STORED instead of ZIP_DEFLATED reduces test runtime from 54 secs to 10 secs

--- a/test/pytest/test_example_torchrec_dlrm.py
+++ b/test/pytest/test_example_torchrec_dlrm.py
@@ -4,13 +4,13 @@ Unit test for the TorchRec DLRM example
 import json
 import shutil
 import sys
-from argparse import Namespace
 from pathlib import Path
 
 import pytest
 import requests
 import test_utils
 import torch
+from model_archiver import ModelArchiverConfig
 
 from ts.torch_handler.unit_tests.test_utils.mock_context import MockContext
 
@@ -90,7 +90,7 @@ def create_mar_file(work_dir, session_mocker, serialized_file, model_archiver):
 
     mar_file_path = Path(work_dir).joinpath(model_name + ".mar")
 
-    args = Namespace(
+    config = ModelArchiverConfig(
         model_name=model_name,
         version="1.0",
         serialized_file=str(serialized_file),
@@ -107,10 +107,8 @@ def create_mar_file(work_dir, session_mocker, serialized_file, model_archiver):
         config_file=None,
     )
 
-    mock = session_mocker.MagicMock()
-    mock.parse_args = session_mocker.MagicMock(return_value=args)
     session_mocker.patch(
-        "archiver.ArgParser.export_model_args_parser", return_value=mock
+        "archiver.ArgParser.export_model_args_parser", return_value=config
     )
 
     # Using ZIP_STORED instead of ZIP_DEFLATED reduces test runtime from 54 secs to 10 secs

--- a/test/pytest/test_parallelism.py
+++ b/test/pytest/test_parallelism.py
@@ -1,13 +1,13 @@
 import json
 import platform
 import shutil
-from argparse import Namespace
 from pathlib import Path
-from unittest.mock import MagicMock, patch
+from unittest.mock import patch
 
 import pytest
 import requests
 import test_utils
+from model_archiver import ModelArchiverConfig
 
 CURR_FILE_PATH = Path(__file__).parent
 REPO_ROOT_DIR = CURR_FILE_PATH.parent.parent
@@ -77,7 +77,7 @@ def create_mar_file(work_dir, model_archiver, model_name):
     handler_py_file = work_dir / "handler.py"
     handler_py_file.write_text(HANDLER_PY)
 
-    args = Namespace(
+    config = ModelArchiverConfig(
         model_name=model_name,
         version="1.0",
         serialized_file=None,
@@ -92,9 +92,7 @@ def create_mar_file(work_dir, model_archiver, model_name):
         config_file=model_config_yaml_file.as_posix(),
     )
 
-    mock = MagicMock()
-    mock.parse_args = MagicMock(return_value=args)
-    with patch("archiver.ArgParser.export_model_args_parser", return_value=mock):
+    with patch("archiver.ArgParser.export_model_args_parser", return_value=config):
         model_archiver.generate_model_archive()
 
         assert mar_file_path.exists()

